File: src/系统/04．伤害系统/02．DOT定义/06．DOT执行器.js
// ========== 虚拟分区：执行器工厂 ==========
export function createDotExecutor(deps) {
    // ========== 虚拟分区：内部状态 ==========
    const EFFECT_RECYCLE_INTERVAL = 0.2;
    const effectRecycleList = [];
    let effectRecycleTimer = undefined;
    let dotTimer = undefined;
    let dotTickBatchTargetHids = null;
    let dotBatchSnapForClear = null;
    let dotBatchDeferredRemaining = 0;
    // ========== 虚拟分区：特效回收 ==========
    function addDotEffectOnUnit(unit, model, duration) {
        if (!unit || !model || model === "" || typeof deps.jass.AddSpecialEffectTarget !== "function")
            return;
        const eff = deps.jass.AddSpecialEffectTarget(model, unit, "origin");
        if (eff == null)
            return;
        if (typeof deps.jass.YDWETimerDestroyEffect === "function") {
            deps.jass.YDWETimerDestroyEffect(duration, eff);
            return;
        }
        const ticks = Math.ceil(duration / EFFECT_RECYCLE_INTERVAL);
        effectRecycleList.push({ eff, ticksLeft: ticks });
        if (effectRecycleTimer == null && typeof deps.jass.TimerStart === "function") {
            effectRecycleTimer = deps.LeakWatcher.createTimer("dot_effectRecycle");
            deps.jass.TimerStart(effectRecycleTimer, EFFECT_RECYCLE_INTERVAL, true, () => {
                for (let i = effectRecycleList.length - 1; i >= 0; i--) {
                    const x = effectRecycleList[i];
                    x.ticksLeft = x.ticksLeft - 1;
                    if (x.ticksLeft <= 0) {
                        if (x.eff != null && typeof deps.jass.DestroyEffect === "function")
                            deps.jass.DestroyEffect(x.eff);
                        effectRecycleList.splice(i, 1);
                    }
                }
                if (effectRecycleList.length === 0 && effectRecycleTimer != null) {
                    deps.LeakWatcher.destroyTimer(effectRecycleTimer);
                    effectRecycleTimer = undefined;
                }
            });
        }
    }
    // ========== 虚拟分区：造成 DOT 伤害 ==========
    function dealDamageForType(typeId, source, target, amount) {
        if (typeof deps.jass.UnitDamageTarget !== "function")
            return;
        const cfg = deps.dotTypes.find(c => c.id === typeId);
        if (cfg == null)
            return;
        if (deps.jass.udg_TempUnit != null) {
            deps.jass.udg_TempUnit[3] = target;
            deps.jass.udg_TempUnit[4] = source;
        }
        const dh = deps.unitHid(target);
        for (let di = 0; di < deps.dotTypes.length; di++) {
            const tid = deps.dotTypes[di].id;
            if (deps.ignoredTargetByType[tid] == null)
                deps.ignoredTargetByType[tid] = {};
            deps.ignoredTargetByType[tid][dh] = true;
        }
        if (typeof deps.damageEventModule.markNextPendingDamageAsDotTickBatch === "function") {
            deps.damageEventModule.markNextPendingDamageAsDotTickBatch();
        }
        deps.jass.UnitDamageTarget(source, target, amount, false, false, deps.jass.ATTACK_TYPE_NORMAL, cfg.damageType, deps.jass.WEAPON_TYPE_WHOKNOWS);
    }
    // ========== 虚拟分区：每秒 tick 执行 ==========
    function dotTickRun() {
        const buffM = require("系统.05．Buff系统.00．Buff系统");
        for (let i = deps.dotTicks.length - 1; i >= 0; i--) {
            const e = deps.dotTicks[i];
            const eh = deps.unitHid(e.target);
            const bid = buffM.DOT_TYPE_TO_BUFF_ID != null ? buffM.DOT_TYPE_TO_BUFF_ID[e.typeId] : undefined;
            const rt = bid != null && bid !== "" && typeof buffM.getBuffRuntimeByHid === "function" ? buffM.getBuffRuntimeByHid(eh, bid) : null;
            if (rt == null || rt.remaining <= 0.001)
                deps.dotTicks.splice(i, 1);
        }
        const batch = {};
        for (let bi = deps.dotTicks.length - 1; bi >= 0; bi--) {
            const bh = deps.unitHid(deps.dotTicks[bi].target);
            if (bh !== 0)
                batch[bh] = true;
        }
        const batchSnap = batch;
        dotTickBatchTargetHids = batchSnap;
        const nDeals = deps.dotTicks.length;
        dotBatchSnapForClear = batchSnap;
        dotBatchDeferredRemaining = nDeals;
        for (let i = deps.dotTicks.length - 1; i >= 0; i--) {
            const e = deps.dotTicks[i];
            const eh = deps.unitHid(e.target);
            dealDamageForType(e.typeId, e.source, e.target, e.amount);
            addDotEffectOnUnit(e.target, e.effectModel, e.effectDuration);
            const cfg = deps.dotTypes.find(c => c.id === e.typeId);
            const stTab = deps.stateByType[e.typeId];
            const stateRaw = stTab != null ? deps.tabRowForHid(stTab, eh) ?? stTab[e.target] : null;
            const state = deps.isValidDotStateRow(stateRaw) ? stateRaw : null;
            if (cfg != null && typeof cfg.onTick === "function" && state != null)
                cfg.onTick(e.target, state);
        }
        if (nDeals <= 0) {
            dotTickBatchTargetHids = null;
            dotBatchSnapForClear = null;
            dotBatchDeferredRemaining = 0;
        }
        if (deps.dotTicks.length === 0 && dotTimer != null) {
            deps.LeakWatcher.destroyTimer(dotTimer);
            dotTimer = undefined;
        }
    }
    // ========== 虚拟分区：计时器保障 ==========
    function ensureDotTimers() {
        if (dotTimer == null && typeof deps.jass.TimerStart === "function") {
            dotTimer = deps.LeakWatcher.createTimer("dot_tick");
            deps.jass.TimerStart(dotTimer, 1, true, dotTickRun);
        }
    }
    // ========== 虚拟分区：批次清理通知 ==========
    function notifyDotTickBatchDamageDisplayed() {
        if (dotBatchDeferredRemaining <= 0)
            return;
        dotBatchDeferredRemaining -= 1;
        if (dotBatchDeferredRemaining <= 0) {
            if (dotTickBatchTargetHids != null && dotTickBatchTargetHids === dotBatchSnapForClear)
                dotTickBatchTargetHids = null;
            dotBatchSnapForClear = null;
            dotBatchDeferredRemaining = 0;
        }
    }
    // ========== 虚拟分区：对外读 batch ==========
    function getDotTickBatchTargetHids() {
        return dotTickBatchTargetHids;
    }
    return {
        ensureDotTimers,
        dealDamageForType,
        notifyDotTickBatchDamageDisplayed,
        getDotTickBatchTargetHids,
    };
}

File: src/系统/04．伤害系统/02．DOT定义/06．DOT执行器.lua
local ____lualib = require("lualib_bundle")
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local ____exports = {}
function ____exports.createDotExecutor(self, deps)
    local EFFECT_RECYCLE_INTERVAL = 0.2
    local effectRecycleList = {}
    local effectRecycleTimer = nil
    local dotTimer = nil
    local dotTickBatchTargetHids = nil
    local dotBatchSnapForClear = nil
    local dotBatchDeferredRemaining = 0
    local function addDotEffectOnUnit(self, unit, model, duration)
        if not unit or not model or model == "" or type(deps.jass.AddSpecialEffectTarget) ~= "function" then
            return
        end
        local eff = deps.jass.AddSpecialEffectTarget(model, unit, "origin")
        if eff == nil then
            return
        end
        if type(deps.jass.YDWETimerDestroyEffect) == "function" then
            deps.jass.YDWETimerDestroyEffect(duration, eff)
            return
        end
        local ticks = math.ceil(duration / EFFECT_RECYCLE_INTERVAL)
        effectRecycleList[#effectRecycleList + 1] = {eff = eff, ticksLeft = ticks}
        if effectRecycleTimer == nil and type(deps.jass.TimerStart) == "function" then
            effectRecycleTimer = deps.LeakWatcher:createTimer("dot_effectRecycle")
            deps.jass.TimerStart(
                effectRecycleTimer,
                EFFECT_RECYCLE_INTERVAL,
                true,
                function()
                    do
                        local i = #effectRecycleList - 1
                        while i >= 0 do
                            local x = effectRecycleList[i + 1]
                            x.ticksLeft = x.ticksLeft - 1
                            if x.ticksLeft <= 0 then
                                if x.eff ~= nil and type(deps.jass.DestroyEffect) == "function" then
                                    deps.jass.DestroyEffect(x.eff)
                                end
                                __TS__ArraySplice(effectRecycleList, i, 1)
                            end
                            i = i - 1
                        end
                    end
                    if #effectRecycleList == 0 and effectRecycleTimer ~= nil then
                        deps.LeakWatcher:destroyTimer(effectRecycleTimer)
                        effectRecycleTimer = nil
                    end
                end
            )
        end
    end
    local function dealDamageForType(self, typeId, source, target, amount)
        if type(deps.jass.UnitDamageTarget) ~= "function" then
            return
        end
        local cfg = __TS__ArrayFind(
            deps.dotTypes,
            function(____, c) return c.id == typeId end
        )
        if cfg == nil then
            return
        end
        local dh = deps:unitHid(target)
        do
            local di = 0
            while di < #deps.dotTypes do
                local tid = deps.dotTypes[di + 1].id
                if deps.ignoredTargetByType[tid] == nil then
                    deps.ignoredTargetByType[tid] = {}
                end
                deps.ignoredTargetByType[tid][dh] = true
                di = di + 1
            end
        end
        if type(deps.damageEventModule.markNextPendingDamageAsDotTickBatch) == "function" then
            deps.damageEventModule:markNextPendingDamageAsDotTickBatch()
        end
        deps.jass.UnitDamageTarget(
            source,
            target,
            amount,
            false,
            false,
            deps.jass.ATTACK_TYPE_NORMAL,
            cfg.damageType,
            deps.jass.WEAPON_TYPE_WHOKNOWS
        )
    end
    local function dotTickRun(self)
        local buffM = require("系统.05．Buff系统.00．Buff系统")
        do
            local i = #deps.dotTicks - 1
            while i >= 0 do
                local e = deps.dotTicks[i + 1]
                local eh = deps:unitHid(e.target)
                local ____temp_0
                if buffM.DOT_TYPE_TO_BUFF_ID ~= nil then
                    ____temp_0 = buffM.DOT_TYPE_TO_BUFF_ID[e.typeId]
                else
                    ____temp_0 = nil
                end
                local bid = ____temp_0
                local ____temp_1
                if bid ~= nil and bid ~= "" and type(buffM.getBuffRuntimeByHid) == "function" then
                    ____temp_1 = buffM:getBuffRuntimeByHid(eh, bid)
                else
                    ____temp_1 = nil
                end
                local rt = ____temp_1
                if rt == nil or rt.remaining <= 0.001 then
                    __TS__ArraySplice(deps.dotTicks, i, 1)
                end
                i = i - 1
            end
        end
        local batch = {}
        do
            local bi = #deps.dotTicks - 1
            while bi >= 0 do
                local bh = deps:unitHid(deps.dotTicks[bi + 1].target)
                if bh ~= 0 then
                    batch[bh] = true
                end
                bi = bi - 1
            end
        end
        local batchSnap = batch
        dotTickBatchTargetHids = batchSnap
        local nDeals = #deps.dotTicks
        dotBatchSnapForClear = batchSnap
        dotBatchDeferredRemaining = nDeals
        do
            local i = #deps.dotTicks - 1
            while i >= 0 do
                local e = deps.dotTicks[i + 1]
                local eh = deps:unitHid(e.target)
                dealDamageForType(
                    nil,
                    e.typeId,
                    e.source,
                    e.target,
                    e.amount
                )
                addDotEffectOnUnit(nil, e.target, e.effectModel, e.effectDuration)
                local cfg = __TS__ArrayFind(
                    deps.dotTypes,
                    function(____, c) return c.id == e.typeId end
                )
                local stTab = deps.stateByType[e.typeId]
                local ____temp_3
                if stTab ~= nil then
                    local ____temp_2 = deps:tabRowForHid(stTab, eh)
                    if ____temp_2 == nil then
                        ____temp_2 = stTab[e.target]
                    end
                    ____temp_3 = ____temp_2
                else
                    ____temp_3 = nil
                end
                local stateRaw = ____temp_3
                local state = deps:isValidDotStateRow(stateRaw) and stateRaw or nil
                if cfg ~= nil and type(cfg.onTick) == "function" and state ~= nil then
                    cfg:onTick(e.target, state)
                end
                i = i - 1
            end
        end
        if nDeals <= 0 then
            dotTickBatchTargetHids = nil
            dotBatchSnapForClear = nil
            dotBatchDeferredRemaining = 0
        end
        if #deps.dotTicks == 0 and dotTimer ~= nil then
            deps.LeakWatcher:destroyTimer(dotTimer)
            dotTimer = nil
        end
    end
    local _registeredToCenterTimer = false
    local function ensureDotTimers(self)
        if _registeredToCenterTimer then
            return
        end
        _registeredToCenterTimer = true
        local ____require_result_4 = require("系统.00．核心系统.05．中心计时器")
        local onSecond = ____require_result_4.onSecond
        onSecond(nil, dotTickRun)
    end
    local function notifyDotTickBatchDamageDisplayed(self)
        if dotBatchDeferredRemaining <= 0 then
            return
        end
        dotBatchDeferredRemaining = dotBatchDeferredRemaining - 1
        if dotBatchDeferredRemaining <= 0 then
            if dotTickBatchTargetHids ~= nil and dotTickBatchTargetHids == dotBatchSnapForClear then
                dotTickBatchTargetHids = nil
            end
            dotBatchSnapForClear = nil
            dotBatchDeferredRemaining = 0
        end
    end
    local function getDotTickBatchTargetHids(self)
        return dotTickBatchTargetHids
    end
    return {ensureDotTimers = ensureDotTimers, dealDamageForType = dealDamageForType, notifyDotTickBatchDamageDisplayed = notifyDotTickBatchDamageDisplayed, getDotTickBatchTargetHids = getDotTickBatchTargetHids}
end
return ____exports
